### PR TITLE
Lint select from `Unit`, or `Int` that incurs widening

### LINKED ISSUE #12728: Lint selections from unit value
## Reproduction steps

Scala version: 2.13.10

```scala
scala> ().
!=(          (universal)   ==(          (universal)   formatted(  (deprecated)   toString()   (universal)
##           (universal)   asInstanceOf (universal)   getClass()                 →(          (deprecated)
+(          (deprecated)   ensuring(    (universal)   hashCode()   (universal)
->(          (universal)   equals(      (universal)   isInstanceOf (universal)
```

## Problem

None of these is a good idea.

https://github.com/scala/scala/pull/6535#issuecomment-383485026
https://github.com/scala/scala/pull/10293#issuecomment-1434095201

Though the parser now supports `() == ()`!

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12728

### CODE REVIEW

**Comment by som-snytt** [comment]
Some current inconsistency.
```
scala> 42.ceil
val res0: Float = 42.0

scala> def i = 42
def i: Int

scala> i.ceil
       ^
       warning: Widening conversion from Int to Float is deprecated because it loses precision. Write `.toFloat` instead.
val res1: Float = 42.0

scala> def c = 'c'
def c: Char

scala> c.ceil
val res2: Float = 99.0
```
where `c.ceil` is pronounced "Cécile" and is even less likely to be intended.

With the new warning, the current warning is less correct.

**Comment by som-snytt** [comment]
It warns in two spots that specialize `Array[Unit]`.

The other thought is that (as with the current warning), any promotion is dubious, it doesn't matter which member is selected.

Then just need `definitions.RichFloatClass`.

**Comment by som-snytt** [comment]
The check is only at typer, which gives a pass to
```
case Literal(Constant(())) =>
```
Not sure why it was deemed bad to allow `case _: () =>` as the unit value ought to be a singleton. (It is not `Unit with Singleton`.) I guess it doesn't matter because the test is always boxed unit equality. There's no advantage to `case _: Unit`.

Something became `().asInstanceOf[Unit]` in fields, a private lazy val after init. (That was the `test execution will be non-parallel` in `PartestDefaults`.)

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -5279,16 +5279,26 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
 
           val sym = tree.symbol orElse member(qual.tpe, name) orElse inCompanionForJavaStatic(qual.tpe.typeSymbol, name)
           if ((sym eq NoSymbol) && name != nme.CONSTRUCTOR && mode.inAny(EXPRmode | PATTERNmode)) {
             // symbol not found? --> try to convert implicitly to a type that does have the required
             // member.  Added `| PATTERNmode` to allow enrichment in patterns (so we can add e.g., an
             // xml member to StringContext, which in turn has an unapply[Seq] method)
 
+              def checkDubiousAdaptation(sel: Tree): Unit = if (!isPastTyper && settings.lintNumericMethods) {
+                val dubious = ScalaIntegralValueClasses(qualTp.typeSymbol) && (
+                  sel.symbol.owner.eq(BoxedFloatClass) || sel.symbol.owner.eq(RichFloatClass))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:5307]: `long.isNaN` goes through `float2Float(long.toFloat)`, not `double2Double`, so this seems to cover it all.

**Review by lrytz — APPROVED** [review]
LGTM, and very useful.

**Comment by som-snytt** [comment]
Rebased, and enhanced `partest --realeasy` to skip tests that run on today's JDK.

```
-- 1 - run/blank.scala                           [skipped on Java 19, compiling against JDK8 but must run on 11+]
```

It would be ideal if `--realeasy` did not skip in case of explicit `scalac: --release:8` but that was too subtle for me.

The option also adds `run/t6240-universe-code-gen.scala` which is annoying to forget to check. (I had actually fixed it in the wrong commit.)
```
Selected 1810 tests drawn from 1 named test categories, 2 tests modified on this branch, and other tests you might have forgotten
```
In case of selection failure,
```
sbt:root> partest junk
Discarding 1 invalid test paths

Selected 0 tests drawn from the well of despair. I see you're not in a testing mood.
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
```

**Review by lrytz — APPROVED** [review]
Thank you!

**Pull request merged**